### ISSUE TITLE
Bug: 5.27.0 regression: @typescript-eslint/no-unused-vars wrongly reports the jsxFactory set in tsconfig.json

### ISSUE THREAD

**Issue by lydell** [typescript-eslint/typescript-eslint#5153]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Issue Description

I use Preact and have this in my tsconfig.json:

```
    "jsx": "react",
    "jsxFactory": "h",
    "jsxFragmentFactory": "Fragment"
```

I use it like so:

```tsx
import { h, Fragment } from "preact";

export default function MyComponent() {
  return <><h1>Hello, world!</h1></>;
}
```

It used to result in no errors. With version 5.27.0, it reports `h` and `Fragment` as unused.

I’ve searched through the issue tracker, and also in TypeScript’s and Preact’s issue trackers, but couldn’t find anything. I apologize if I missed something, or this is a misunderstanding on my side that just happened to work before!

### Reproduction Repository Link

https://github.com/lydell/typescript-eslint-preact-issue

### Repro Steps

1) clone the repo
2) `npm ci`
3) `npm test`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.27.0` |
| `@typescript-eslint/parser`            | `5.27.0` |
| `TypeScript`                           | `4.7.3` |
| `ESLint`                               | `8.17.0` |
| `node`                                 | `18.1.0` |

**Comment by bradzacher** [comment]
I can check later today when I'm at my laptop, but I think all you need is to set `parserOptions.jsx = true`

We probably should remove that restriction though as we don't use that anywhere else.

**Comment by lydell** [comment]
I can confirm that adding this to my ESLint config makes it work:

```
"parserOptions": {
  "ecmaFeatures": {
    "jsx": true
  }
}
```

**Comment by bradzacher** [comment]
For this issue we want to remove the requirement for the `parserOptions.ecmaFeatures.jsx` option to be set for the automated pragma resolution. Should be a simple change!

**Comment by Solo-steven** [comment]
Hi @bradzacher , I am new there and I want to take this issue. After some research, I find two solutions,
The first is to make jsx option fallback to true, which would make the parser automatically resolve jsx pragma.
<img width="434" alt="Screen Shot 2022-11-25 at 6 52 06 PM" src="https://user-images.githubusercontent.com/67428009/203967740-cad5b57f-fb5d-4f84-b979-2d138e882224.png">
<img width="603" alt="Screen Shot 2022-11-25 at 6 53 20 PM" src="https://user-images.githubusercontent.com/67428009/203968040-9e736169-88c7-4904-820c-b50293468c28.png">

The second solution is to emit the if statement before pragma resolves, but this method would make the parser resolve jsx pragma even if jsx option is false. So I think the first solution is a better one. 
<img width="603" alt="Screen Shot 2022-11-25 at 6 54 13 PM" src="https://user-images.githubusercontent.com/67428009/203968217-ae9149ef-78db-42e4-9abc-0a96bc67a570.png">

looking forward to your response!

**Comment by bradzacher** [comment]
Given that JSX pragma and factory are only used by the tooling if we encounter JSX syntax in the code - there's not really any cost to us resolving the config even if `jsx == false`.

In fact it's better if we do that because in a lot of cases we actually ignore the `jsx` setting!

So let's go with the second solution of just removing the `parserOptions.jsx` check entirely.
